File: zxing-lite/src/main/java/com/king/zxing/DecodeFormatManager.java
package com.king.zxing;

import com.google.zxing.BarcodeFormat;
import com.google.zxing.DecodeHintType;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;

import androidx.annotation.NonNull;

/**
 * 解码格式管理器
 * <p>
 * 将常见的一些解码配置已根据条形码类型进行了几大划分，可根据需要找到符合的划分配置类型直接使用。
 *
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 */
public final class DecodeFormatManager {

    /**
     * 所有的
     */
    public static final Map<DecodeHintType, Object> ALL_HINTS = new EnumMap<>(DecodeHintType.class);
    /**
     * CODE_128 (最常用的一维码)
     */
    public static final Map<DecodeHintType, Object> CODE_128_HINTS = createDecodeHint(BarcodeFormat.CODE_128);
    /**
     * QR_CODE (最常用的二维码)
     */
    public static final Map<DecodeHintType, Object> QR_CODE_HINTS = createDecodeHint(BarcodeFormat.QR_CODE);
    /**
     * 一维码
     */
    public static final Map<DecodeHintType, Object> ONE_DIMENSIONAL_HINTS = new EnumMap<>(DecodeHintType.class);
    /**
     * 二维码
     */
    public static final Map<DecodeHintType, Object> TWO_DIMENSIONAL_HINTS = new EnumMap<>(DecodeHintType.class);
    /**
     * 默认
     */
    public static final Map<DecodeHintType, Object> DEFAULT_HINTS = new EnumMap<>(DecodeHintType.class);

    static {
        //all hints
        addDecodeHintTypes(ALL_HINTS, getAllFormats());
        //one dimension
        addDecodeHintTypes(ONE_DIMENSIONAL_HINTS, getOneDimensionalFormats());
        //Two dimension
        addDecodeHintTypes(TWO_DIMENSIONAL_HINTS, getTwoDimensionalFormats());
        //default hints
        addDecodeHintTypes(DEFAULT_HINTS, getDefaultFormats());
    }

    /**
     * 所有支持的{@link BarcodeFormat}
     *
     * @return
     */
    private static List<BarcodeFormat> getAllFormats() {
        List<BarcodeFormat> list = new ArrayList<>();
        list.add(BarcodeFormat.AZTEC);
        list.add(BarcodeFormat.CODABAR);
        list.add(BarcodeFormat.CODE_39);
        list.add(BarcodeFormat.CODE_93);
        list.add(BarcodeFormat.CODE_128);
        list.add(BarcodeFormat.DATA_MATRIX);
        list.add(BarcodeFormat.EAN_8);
        list.add(BarcodeFormat.EAN_13);
        list.add(BarcodeFormat.ITF);
        list.add(BarcodeFormat.MAXICODE);
        list.add(BarcodeFormat.PDF_417);
        list.add(BarcodeFormat.QR_CODE);
        list.add(BarcodeFormat.RSS_14);
        list.add(BarcodeFormat.RSS_EXPANDED);
        list.add(BarcodeFormat.UPC_A);
        list.add(BarcodeFormat.UPC_E);
        list.add(BarcodeFormat.UPC_EAN_EXTENSION);
        return list;
    }

    /**
     * 一维码
     * 包括如下几种格式：
     * {@link BarcodeFormat#CODABAR}
     * {@link BarcodeFormat#CODE_39}
     * {@link BarcodeFormat#CODE_93}
     * {@link BarcodeFormat#CODE_128}
     * {@link BarcodeFormat#EAN_8}
     * {@link BarcodeFormat#EAN_13}
     * {@link BarcodeFormat#ITF}
     * {@link BarcodeFormat#RSS_14}
     * {@link BarcodeFormat#RSS_EXPANDED}
     * {@link BarcodeFormat#UPC_A}
     * {@link BarcodeFormat#UPC_E}
     * {@link BarcodeFormat#UPC_EAN_EXTENSION}
     *
     * @return
     */
    private static List<BarcodeFormat> getOneDimensionalFormats() {
        List<BarcodeFormat> list = new ArrayList<>();
        list.add(BarcodeFormat.CODABAR);
        list.add(BarcodeFormat.CODE_39);
        list.add(BarcodeFormat.CODE_93);
        list.add(BarcodeFormat.CODE_128);
        list.add(BarcodeFormat.EAN_8);
        list.add(BarcodeFormat.EAN_13);
        list.add(BarcodeFormat.ITF);
        list.add(BarcodeFormat.RSS_14);
        list.add(BarcodeFormat.RSS_EXPANDED);
        list.add(BarcodeFormat.UPC_A);
        list.add(BarcodeFormat.UPC_E);
        list.add(BarcodeFormat.UPC_EAN_EXTENSION);
        return list;
    }

    /**
     * 二维码
     * 包括如下几种格式：
     * {@link BarcodeFormat#AZTEC}
     * {@link BarcodeFormat#DATA_MATRIX}
     * {@link BarcodeFormat#MAXICODE}
     * {@link BarcodeFormat#PDF_417}
     * {@link BarcodeFormat#QR_CODE}
     *
     * @return
     */
    private static List<BarcodeFormat> getTwoDimensionalFormats() {
        List<BarcodeFormat> list = new ArrayList<>();
        list.add(BarcodeFormat.AZTEC);
        list.add(BarcodeFormat.DATA_MATRIX);
        list.add(BarcodeFormat.MAXICODE);
        list.add(BarcodeFormat.PDF_417);
        list.add(BarcodeFormat.QR_CODE);
        return list;
    }

    /**
     * 默认支持的格式
     * 包括如下几种格式：
     * {@link BarcodeFormat#QR_CODE}
     * {@link BarcodeFormat#UPC_A}
     * {@link BarcodeFormat#EAN_13}
     * {@link BarcodeFormat#CODE_128}
     *
     * @return
     */
    private static List<BarcodeFormat> getDefaultFormats() {
        List<BarcodeFormat> list = new ArrayList<>();
        list.add(BarcodeFormat.QR_CODE);
        list.add(BarcodeFormat.UPC_A);
        list.add(BarcodeFormat.EAN_13);
        list.add(BarcodeFormat.CODE_128);
        return list;
    }

    /**
     * 支持解码的格式
     *
     * @param barcodeFormats {@link BarcodeFormat}
     * @return
     */
    public static Map<DecodeHintType, Object> createDecodeHints(@NonNull BarcodeFormat... barcodeFormats) {
        Map<DecodeHintType, Object> hints = new EnumMap<>(DecodeHintType.class);
        addDecodeHintTypes(hints, Arrays.asList(barcodeFormats));
        return hints;
    }

    /**
     * 支持解码的格式
     *
     * @param barcodeFormat {@link BarcodeFormat}
     * @return
     */
    public static Map<DecodeHintType, Object> createDecodeHint(@NonNull BarcodeFormat barcodeFormat) {
        Map<DecodeHintType, Object> hints = new EnumMap<>(DecodeHintType.class);
        addDecodeHintTypes(hints, Collections.singletonList(barcodeFormat));
        return hints;
    }

    /**
     * @param hints
     * @param formats
     */
    private static void addDecodeHintTypes(Map<DecodeHintType, Object> hints, List<BarcodeFormat> formats) {
        // Image is known to be of one of a few possible formats.
        hints.put(DecodeHintType.POSSIBLE_FORMATS, formats);
        // Spend more time to try to find a barcode; optimize for accuracy, not speed.
        hints.put(DecodeHintType.TRY_HARDER, Boolean.TRUE);
        // Specifies what character encoding to use when decoding, where applicable (type String)
        hints.put(DecodeHintType.CHARACTER_SET, "UTF-8");
    }

}

File: zxing-lite/src/main/java/com/king/zxing/DecodeFormatManager.kt
package com.king.zxing

import androidx.annotation.NonNull
import com.google.zxing.BarcodeFormat
import com.google.zxing.DecodeHintType
import java.util.ArrayList
import java.util.Arrays
import java.util.Collections
import java.util.EnumMap

/**
 * 解码格式管理器
 * <p>
 * 将常见的一些解码配置已根据条形码类型进行了几大划分，可根据需要找到符合的划分配置类型直接使用。
 *
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 * <p>
 * <a href="https://github.com/jenly1314">Follow me</a>
 */
object DecodeFormatManager {

    @JvmField
    val ALL_HINTS: MutableMap<DecodeHintType, Any> = EnumMap(DecodeHintType::class.java)

    @JvmField
    val CODE_128_HINTS: MutableMap<DecodeHintType, Any> = createDecodeHint(BarcodeFormat.CODE_128)

    @JvmField
    val QR_CODE_HINTS: MutableMap<DecodeHintType, Any> = createDecodeHint(BarcodeFormat.QR_CODE)

    @JvmField
    val ONE_DIMENSIONAL_HINTS: MutableMap<DecodeHintType, Any> = EnumMap(DecodeHintType::class.java)

    @JvmField
    val TWO_DIMENSIONAL_HINTS: MutableMap<DecodeHintType, Any> = EnumMap(DecodeHintType::class.java)

    @JvmField
    val DEFAULT_HINTS: MutableMap<DecodeHintType, Any> = EnumMap(DecodeHintType::class.java)

    init {
        addDecodeHintTypes(ALL_HINTS, getAllFormats())
        addDecodeHintTypes(ONE_DIMENSIONAL_HINTS, getOneDimensionalFormats())
        addDecodeHintTypes(TWO_DIMENSIONAL_HINTS, getTwoDimensionalFormats())
        addDecodeHintTypes(DEFAULT_HINTS, getDefaultFormats())
    }

    private fun getAllFormats(): List<BarcodeFormat> {
        return ArrayList<BarcodeFormat>().apply {
            add(BarcodeFormat.AZTEC)
            add(BarcodeFormat.CODABAR)
            add(BarcodeFormat.CODE_39)
            add(BarcodeFormat.CODE_93)
            add(BarcodeFormat.CODE_128)
            add(BarcodeFormat.DATA_MATRIX)
            add(BarcodeFormat.EAN_8)
            add(BarcodeFormat.EAN_13)
            add(BarcodeFormat.ITF)
            add(BarcodeFormat.MAXICODE)
            add(BarcodeFormat.PDF_417)
            add(BarcodeFormat.QR_CODE)
            add(BarcodeFormat.RSS_14)
            add(BarcodeFormat.RSS_EXPANDED)
            add(BarcodeFormat.UPC_A)
            add(BarcodeFormat.UPC_E)
            add(BarcodeFormat.UPC_EAN_EXTENSION)
        }
    }

    private fun getOneDimensionalFormats(): List<BarcodeFormat> {
        return ArrayList<BarcodeFormat>().apply {
            add(BarcodeFormat.CODABAR)
            add(BarcodeFormat.CODE_39)
            add(BarcodeFormat.CODE_93)
            add(BarcodeFormat.CODE_128)
            add(BarcodeFormat.EAN_8)
            add(BarcodeFormat.EAN_13)
            add(BarcodeFormat.ITF)
            add(BarcodeFormat.RSS_14)
            add(BarcodeFormat.RSS_EXPANDED)
            add(BarcodeFormat.UPC_A)
            add(BarcodeFormat.UPC_E)
            add(BarcodeFormat.UPC_EAN_EXTENSION)
        }
    }

    private fun getTwoDimensionalFormats(): List<BarcodeFormat> {
        return ArrayList<BarcodeFormat>().apply {
            add(BarcodeFormat.AZTEC)
            add(BarcodeFormat.DATA_MATRIX)
            add(BarcodeFormat.MAXICODE)
            add(BarcodeFormat.PDF_417)
            add(BarcodeFormat.QR_CODE)
        }
    }

    private fun getDefaultFormats(): List<BarcodeFormat> {
        return ArrayList<BarcodeFormat>().apply {
            add(BarcodeFormat.QR_CODE)
            add(BarcodeFormat.UPC_A)
            add(BarcodeFormat.EAN_13)
            add(BarcodeFormat.CODE_128)
        }
    }

    @JvmStatic
    fun createDecodeHints(@NonNull vararg barcodeFormats: BarcodeFormat): MutableMap<DecodeHintType, Any> {
        val hints: MutableMap<DecodeHintType, Any> = EnumMap(DecodeHintType::class.java)
        addDecodeHintTypes(hints, Arrays.asList(*barcodeFormats))
        return hints
    }

    @JvmStatic
    fun createDecodeHint(@NonNull barcodeFormat: BarcodeFormat): MutableMap<DecodeHintType, Any> {
        val hints: MutableMap<DecodeHintType, Any> = EnumMap(DecodeHintType::class.java)
        addDecodeHintTypes(hints, Collections.singletonList(barcodeFormat))
        return hints
    }

    private fun addDecodeHintTypes(hints: MutableMap<DecodeHintType, Any>, formats: List<BarcodeFormat>) {
        hints[DecodeHintType.POSSIBLE_FORMATS] = formats
        hints[DecodeHintType.TRY_HARDER] = true
        hints[DecodeHintType.CHARACTER_SET] = "UTF-8"
    }
}
